src/executables/srpd_aging.c

```diff
@@ -0,0 +1,332 @@
+#include <sys/wait.h>
+
+#include <libyang/libyang.h>
+#include <sysrepo.h>
+
+#include "srpd_common.h"
+#include "srpd_aging.h"
+#include "config.h"
+#include "bin_common.h"
+
+char *
+get_notif_path(void)
+{
+	char *path;
+	if (SR_NOTIFICATION_PATH[0]) {
+		path = strdup(SR_NOTIFICATION_PATH);
+	} else {
+		if (asprintf(&path, "%s/data/notif/", sr_get_repo_path()) == -1) {
+			path = NULL;
+		}
+	}
+	return path; 
+}
+
+int
+sr_my_state_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_name, const char *path,
+		const char *request_xpath, uint32_t request_id, struct lyd_node **parent, void *private_data)
+{
+	int rc = 0;
+	const struct ly_ctx *ctx;
+	char value[20];
+	dir_st_t *opts = (dir_st_t *)private_data;
+
+	(void) sub_id;
+	(void) module_name;
+	(void) path;
+	(void) request_xpath;
+	(void) request_id;
+
+	sprintf(value, "%" PRIu64, opts->archived_notif_count);
+	ctx = sr_session_acquire_context(session);
+	if ((rc = lyd_new_path(*parent, ctx, "/sysrepo-plugind:sysrepo-plugind/notif-datastore/rotation/archived-files-count",
+			value, 0, NULL)) != LY_SUCCESS) {
+				goto cleanup;
+	}
+
+cleanup:
+	sr_session_release_context(session);
+	return rc;
+}
+
+int
+format_control(char *string, time_t *file_time1, time_t *file_time2)
+{
+	char *x = NULL, *endptr = NULL;
+	time_t time1, time2;
+	char buf[7];
+
+	memset(buf, '\0', 7);
+
+	if (string == NULL) {
+		return EXIT_FAILURE;
+	}
+	x = strchr(string, '.');
+	if (x == NULL) {
+		return EXIT_FAILURE;
+	}
+	strncpy(buf, x + 1, 6);
+	if (strcmp(buf, "notif.")) {
+		return EXIT_FAILURE;
+	}
+	x = strchr(x + 1, '.');
+	time1 = strtoul(x + 1, &endptr, 10);
+	if (!strcmp(endptr, x + 1)) {
+		return EXIT_FAILURE;
+	}
+	x = strchr(x + 1, '-');
+	if (x == NULL) {
+		return EXIT_FAILURE;
+	}
+	time2 = strtoul(x + 1, &endptr, 10);
+	if (!strcmp(endptr, x + 1)) {
+		return EXIT_FAILURE;
+	}
+
+	if (file_time1 != NULL) {
+		*file_time1 = time1;
+	}
+	if (file_time2 != NULL) {
+		*file_time2 = time2;
+	}
+	return 0;
+}
+
+int
+execute_cmd(const char *cmd, const char *arg1, const char *arg2)
+{
+	pid_t pid;
+	int ret;
+	
+	pid = fork();
+	if (pid == 0) {
+		if (execl(cmd, cmd, arg1, arg2, NULL) == -1) {
+			fprintf(stderr, "Execl failed: %s\n", strerror(errno));
+			exit(1);
+		}
+	} else if (pid == -1) {
+		fprintf(stderr, "Fork failed: %s\n", strerror(errno));
+		return EXIT_FAILURE;
+	}
+	waitpid(pid, &ret, 0);
+	if (WIFSIGNALED(ret)) {
+		fprintf(stderr, "Child has been terminated by a signal no: %d\n", WTERMSIG(ret));
+	}
+	return 0;
+}
+
+void *
+sr_archive_loop(void *arg)
+{
+	int rc = 0;
+	DIR *d = NULL;
+	struct dirent *dir = NULL;
+	time_t current_time = 0, last_time = 0, file_time2 = 0;
+	dir_st_t *opts = (dir_st_t *)arg;
+	char *arg1 = NULL, *arg2 = NULL, *remove_str = NULL, *notif_dir_name = NULL;
+	
+	notif_dir_name = get_notif_path();
+	if (!notif_dir_name) {
+		fprintf(stderr, "Fatal error: Notif directory is NULL\n");
+		goto cleanup;
+	}
+
+	while (opts->running) {
```

Comment:
This looks like a busy wait loop. Why not sleep until it's time to perform an activity, and only then wake up?

Reply:
Fixed